JavaScript/chapter01/p02_check_perm/avc278.js

```diff
@@ -0,0 +1,71 @@
+// Check Permutation: Given two strings, write a method to decide if one is a permutation of the other.
+
+const assert = require("assert");
+
+/**
+ * Checks if a string is a permutation of another
+ * @param  {string} A source input string to check against
+ * @param  {string} B target input string to compare with source string
+ * @return {bool}     Whether the input strings are permutations of one another
+ *
+ * Throughout the problem, we set up a couple ways to early exit. At the start, if both strings are not of the same
+ * length, we already know they cannot be permutations of each other, so we return false.
+ * Otherwise, we can continue and set up a mapping between the elements in A and B, and their frequencies.
+ * This is an O(N) operation with respect to time and space where N is the length of the string A.
+ * At this point, we can get the keys from the frequency objects we created, which is an O(N) operation for space.
+ * If the lengths of `aKeys` and `bKeys` are not the same, we know they cannot be permutations and we exit early.
+ * Otherwise, we can continue to iterate through the keys. We need to make sure that the frequencies of aKeys exist in
+ * both aFreq and bFreq AND that the frequencies of bKeys exist in aFreq and bFreq. It is not enough to check that one
```

Comment:
> It is not enough to check that one exists in both and not check the other set of keys.

Can this statement be elaborated? Is there an example where it is not enough?

Reply:
Hmm yeah, so initially when thinking about this problem, I was thinking about some math class I took on set theory where if you want to prove that set A and set B are identical, you need to show that set A is a subset of B, and B is a subset of A.
So the initial thought process was
```
aFreq = {
  'a': 1,
  'b': 1,
  'c': 2,
}
bFreq = {
  'a': 1,
  'b': 1,
  'c': 2,
  'd': 1,
}
```
If we only check the frequencies of A match with the frequencies of B, then we should be good to go, HOWEVER, now that I'm thinking about it some more, we actually don't need to check that as this type of scenario will never happen since we already have checks in place that will return false early on.

I'm going to remove that part in the next commit